File: mobile/src/services/profiles.ts
/**
 * Local ("online") profiles, ported from `profiles.rs` and the `local_profile_*`
 * commands of `lib.rs`. The list lives at `KEYS.localProfiles`; PINs in SecureStore.
 */
import type { LocalProfile, ProfilePatch, Session } from "../lib/types";
import { KEYS, SECRET, secrets, store } from "./store";
import { nowMs, uuid } from "./util";
import {
  MAX_PROFILES,
  cleanAvatar,
  cleanName,
  cleanPin,
  profileView,
  type StoredLocalProfile,
} from "./profiles.pure";
import * as session from "./session";
import { checkCreate, checkDelete } from "./parental";

type DeleteHook = (profileId: string) => Promise<void> | void;
const deleteHooks = new Set<DeleteHook>();

/** Extra cleanup when a profile is deleted (IPTV files, databases…). */
export function registerProfileDeleteHook(fn: DeleteHook): () => void {
  deleteHooks.add(fn);
  return () => {
    deleteHooks.delete(fn);
  };
}

function view(profile: StoredLocalProfile): LocalProfile {
  return profileView(profile, session.hasLinkedSession(profile.id));
}

function find(id: string): StoredLocalProfile {
  const profile = session.readLocalProfiles().find((p) => p.id === id);
  if (!profile) throw new Error("Perfil no encontrado");
  return profile;
}

export async function localProfilesList(): Promise<LocalProfile[]> {
  return session.readLocalProfiles().map(view);
}

/** True when `pin` is the profile's PIN (or it has none). */
async function pinMatches(profile: StoredLocalProfile, pin: string | null | undefined): Promise<boolean> {
  if (!profile.hasPin) return true;
  const stored = await secrets.get(SECRET.pin(profile.id));
  return pin != null && stored != null && stored === pin;
}

/** Checks a profile's PIN without opening it (editing it from the profile picker). */
export async function localProfileCheckPin(id: string, pin: string): Promise<void> {
  if (!(await pinMatches(find(id), pin))) throw new Error("PIN incorrecto");
}

/** `parentalPin` is asked for while a profile has a parental restriction. */
export async function localProfileCreate(
  name: string,
  avatar: string,
  pin: string | null,
  parentalPin?: string | null,
): Promise<LocalProfile> {
  await checkCreate(parentalPin);
  const profiles = session.readLocalProfiles();
  if (profiles.length >= MAX_PROFILES) throw new Error(`Máximo ${MAX_PROFILES} perfiles`);
  const cleanedPin = cleanPin(pin);
  const profile: StoredLocalProfile = {
    id: uuid(),
    name: cleanName(name),
    avatar: cleanAvatar(avatar),
    hasPin: cleanedPin != null,
    createdMs: nowMs(),
  };
  if (cleanedPin) await secrets.set(SECRET.pin(profile.id), cleanedPin);
  profiles.push(profile);
  session.writeLocalProfiles(profiles);
  return view(profile);
}

export async function localProfileUpdate(id: string, patch: ProfilePatch): Promise<LocalProfile> {
  const profiles = session.readLocalProfiles();
  const profile = profiles.find((p) => p.id === id);
  if (!profile) throw new Error("Perfil no encontrado");
  // Changing or removing the PIN of a profile that is not open takes that PIN.
  const pinChange = patch.clearPin || patch.pin != null;
  if (pinChange && session.activeLocalProfileId() !== id && !(await pinMatches(profile, patch.currentPin))) {
    throw new Error("PIN incorrecto");
  }
  if (patch.name != null) profile.name = cleanName(patch.name);
  if (patch.avatar != null) profile.avatar = cleanAvatar(patch.avatar);
  if (patch.clearPin) {
    await secrets.remove(SECRET.pin(id));
    profile.hasPin = false;
  } else if (patch.pin != null) {
    const cleaned = cleanPin(patch.pin);
    if (cleaned) {
      await secrets.set(SECRET.pin(id), cleaned);
      profile.hasPin = true;
    } else {
      await secrets.remove(SECRET.pin(id));
      profile.hasPin = false;
    }
  }
  session.writeLocalProfiles(profiles);
  if (session.activeLocalProfileId() === id) session.setActiveLocalProfile({ ...profile });
  return view(profile);
}

/** Removes the profile with its settings, addon progress, IPTV data and linked session. */
export async function localProfileDelete(
  id: string,
  pin?: string | null,
  parentalPin?: string | null,
): Promise<void> {
  const profile = session.readLocalProfiles().find((p) => p.id === id);
  if (profile && session.activeLocalProfileId() !== id && !(await pinMatches(profile, pin))) {
    throw new Error("PIN incorrecto");
  }
  await checkDelete(id, parentalPin);
  if (session.activeLocalProfileId() === id) {
    await session.clearSessionCaches();
    session.setActiveLocalProfile(null);
    session.setJellyfinSession(null);
  }
  session.writeLocalProfiles(session.readLocalProfiles().filter((p) => p.id !== id));
  await session.clearLinkedSession(id);
  await secrets.remove(SECRET.pin(id));
  store.remove(KEYS.settings(id));
  store.remove(KEYS.addonProgress(id));
  store.remove(KEYS.iptv(id));
  store.remove(KEYS.iptvFavorites(id));
  store.remove(KEYS.iptvRecent(id));
  for (const hook of Array.from(deleteHooks)) {
    try {
      await hook(id);
    } catch (error) {
      console.warn("[profiles] delete hook failed", error);
    }
  }
  if (store.get<string>(KEYS.activeLocalProfile) === id) store.remove(KEYS.activeLocalProfile);
}

/** Opens a local profile (after checking its PIN) and restores its linked Jellyfin account. */
export async function localProfileEnter(id: string, pin: string | null): Promise<Session> {
  const profile = find(id);
  if (profile.hasPin) {
    const stored = await secrets.get(SECRET.pin(id));
    if (pin == null || stored == null || stored !== pin) throw new Error("PIN incorrecto");
  }
  await session.clearSessionCaches();
  session.setJellyfinSession(null);
  session.setActiveLocalProfile(profile);
  await session.restoreLinkedSession(profile.id);
  const account = session.accountView();
  if (!account) throw new Error("No hay sesión activa");
  return account;
}

File: src-tauri/src/profiles.rs
//! Local ("online") profiles: identities that live only inside the app, for people who
//! watch through Stremio addons without a Jellyfin server. A profile can be protected
//! with a 4-digit PIN and can later link a Jellyfin account (kept per profile).

use serde::{Deserialize, Serialize};
use serde_json::Value;
use tauri_plugin_store::StoreExt;

pub const MAX_PROFILES: usize = 8;
const MAX_NAME: usize = 40;
const MAX_AVATAR_BYTES: usize = 400 * 1024;
const PRESET_COUNT: u32 = 12;
const PROFILES_KEY: &str = "localProfiles";
const ACTIVE_KEY: &str = "activeLocalProfile";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalProfile {
    pub id: String,
    pub name: String,
    /// `preset:<n>` or a `data:image/…;base64,…` picture chosen by the user.
    pub avatar: String,
    /// DPAPI-protected PIN (hex); `None` when the profile opens without one.
    #[serde(default)]
    pub pin: Option<String>,
    #[serde(default)]
    pub created_ms: u64,
}

/// What the frontend sees (never the PIN).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalProfileView {
    pub id: String,
    pub name: String,
    pub avatar: String,
    pub has_pin: bool,
    /// A Jellyfin account is linked to this profile.
    pub linked: bool,
}

/// Partial update sent by the profile editor.
#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct ProfilePatch {
    pub name: Option<String>,
    pub avatar: Option<String>,
    /// New PIN (4 digits). Ignored when `clear_pin` is set.
    pub pin: Option<String>,
    pub clear_pin: bool,
    /// The PIN the profile has now, to change or remove it while it is not open.
    pub current_pin: Option<String>,
}

impl LocalProfile {
    pub fn view(&self, app: &tauri::AppHandle) -> LocalProfileView {
        LocalProfileView {
            id: self.id.clone(),
            name: self.name.clone(),
            avatar: self.avatar.clone(),
            has_pin: self.pin.is_some(),
            linked: has_linked_session(app, &self.id),
        }
    }

    pub fn verify_pin(&self, pin: Option<&str>) -> bool {
        match &self.pin {
            None => true,
            Some(sealed) => {
                let Some(pin) = pin else { return false };
                let Ok(bytes) = crate::protect::from_hex(sealed) else {
                    return false;
                };
                let Ok(raw) = crate::protect::unprotect(&bytes) else {
                    return false;
                };
                raw == pin.as_bytes()
            }
        }
    }
}

/// Store key of the Jellyfin session linked to a local profile.
pub fn session_key(id: &str) -> String {
    format!("localSession.{id}")
}

fn has_linked_session(app: &tauri::AppHandle, id: &str) -> bool {
    app.store(crate::store_path())
        .ok()
        .map(|store| store.get(session_key(id)).is_some())
        .unwrap_or(false)
}

pub fn list(app: &tauri::AppHandle) -> Result<Vec<LocalProfile>, String> {
    let store = app.store(crate::store_path()).map_err(|e| e.to_string())?;
    match store.get(PROFILES_KEY) {
        Some(value) => Ok(serde_json::from_value::<Vec<LocalProfile>>(value).unwrap_or_default()),
        None => Ok(vec![]),
    }
}

fn save_list(app: &tauri::AppHandle, list: &[LocalProfile]) -> Result<(), String> {
    let store = app.store(crate::store_path()).map_err(|e| e.to_string())?;
    store.set(
        PROFILES_KEY,
        serde_json::to_value(list).map_err(|e| e.to_string())?,
    );
    crate::save_store(&store)
}

pub fn get(app: &tauri::AppHandle, id: &str) -> Result<Option<LocalProfile>, String> {
    Ok(list(app)?.into_iter().find(|p| p.id == id))
}

pub fn create(
    app: &tauri::AppHandle,
    name: &str,
    avatar: &str,
    pin: Option<&str>,
) -> Result<LocalProfile, String> {
    let mut profiles = list(app)?;
    if profiles.len() >= MAX_PROFILES {
        return Err(format!("Máximo {MAX_PROFILES} perfiles"));
    }
    let profile = LocalProfile {
        id: uuid::Uuid::new_v4().to_string(),
        name: clean_name(name)?,
        avatar: clean_avatar(avatar)?,
        pin: seal_pin(pin)?,
        created_ms: crate::addons::now_ms(),
    };
    profiles.push(profile.clone());
    save_list(app, &profiles)?;
    Ok(profile)
}

pub fn update(app: &tauri::AppHandle, id: &str, patch: ProfilePatch) -> Result<LocalProfile, String> {
    let mut profiles = list(app)?;
    let profile = profiles
        .iter_mut()
        .find(|p| p.id == id)
        .ok_or_else(|| "Perfil no encontrado".to_string())?;
    if let Some(name) = patch.name {
        profile.name = clean_name(&name)?;
    }
    if let Some(avatar) = patch.avatar {
        profile.avatar = clean_avatar(&avatar)?;
    }
    if patch.clear_pin {
        profile.pin = None;
    } else if let Some(pin) = patch.pin {
        profile.pin = seal_pin(Some(&pin))?;
    }
    let updated = profile.clone();
    save_list(app, &profiles)?;
    Ok(updated)
}

/// Removes the profile with its settings, addon progress and linked session.
pub fn delete(app: &tauri::AppHandle, id: &str) -> Result<(), String> {
    let mut profiles = list(app)?;
    profiles.retain(|p| p.id != id);
    save_list(app, &profiles)?;
    let store = app.store(crate::store_path()).map_err(|e| e.to_string())?;
    store.delete(session_key(id));
    store.delete(format!("settings.{id}"));
    store.delete(format!("addonProgress.{id}"));
    drop(store);
    crate::iptv::delete_profile_data(app, id);
    let store = app.store(crate::store_path()).map_err(|e| e.to_string())?;
    if active_id(app).as_deref() == Some(id) {
        store.delete(ACTIVE_KEY);
    }
    crate::save_store(&store)
}

/// Profile restored on the next launch.
pub fn active_id(app: &tauri::AppHandle) -> Option<String> {
    let store = app.store(crate::store_path()).ok()?;
    store
        .get(ACTIVE_KEY)
        .and_then(|v| v.as_str().map(|s| s.to_string()))
        .filter(|s| !s.is_empty())
}

pub fn set_active(app: &tauri::AppHandle, id: Option<&str>) -> Result<(), String> {
    let store = app.store(crate::store_path()).map_err(|e| e.to_string())?;
    match id {
        Some(id) => store.set(ACTIVE_KEY, Value::String(id.to_string())),
        None => {
            store.delete(ACTIVE_KEY);
        }
    }
    crate::save_store(&store)
}

fn clean_name(name: &str) -> Result<String, String> {
    let name: String = name.trim().chars().take(MAX_NAME).collect();
    if name.is_empty() {
        return Err("Escribe un nombre para el perfil".into());
    }
    Ok(name)
}

fn clean_avatar(avatar: &str) -> Result<String, String> {
    let avatar = avatar.trim();
    if let Some(n) = avatar.strip_prefix("preset:") {
        return match n.parse::<u32>() {
            Ok(n) if n < PRESET_COUNT => Ok(avatar.to_string()),
            _ => Ok("preset:0".to_string()),
        };
    }
    let ok_prefix = avatar.starts_with("data:image/jpeg;base64,")
        || avatar.starts_with("data:image/png;base64,")
        || avatar.starts_with("data:image/webp;base64,");
    if !ok_prefix {
        return Err("Imagen de perfil no válida".into());
    }
    if avatar.len() > MAX_AVATAR_BYTES {
        return Err("La foto es demasiado grande".into());
    }
    if !avatar.bytes().all(|b| b.is_ascii_graphic()) {
        return Err("Imagen de perfil no válida".into());
    }
    Ok(avatar.to_string())
}

fn seal_pin(pin: Option<&str>) -> Result<Option<String>, String> {
    let Some(pin) = pin.map(str::trim).filter(|p| !p.is_empty()) else {
        return Ok(None);
    };
    if pin.len() != 4 || !pin.bytes().all(|b| b.is_ascii_digit()) {
        return Err("El PIN debe tener 4 dígitos".into());
    }
    let sealed = crate::protect::protect(pin.as_bytes())?;
    Ok(Some(crate::protect::to_hex(&sealed)))
}
